### PR TITLE
signup form validation

### LINKED ISSUE #1503: Bug while signing up in the "Name"
Signing up, in the "Name" option.
While signing up where the blank asks to fill the Name.
The blank accepts any random symbol and any amount of unwanted space.
Example: the Name blank accepts "/*--*/+uioyr76$%^^" as name.

**To Reproduce**
1. Go to 'https://circuitverse.org/'
2. Click on 'login/signup'
3. Scroll down to 'signup'
4. In the Name blank when try to enter name as "/*--*/+uioyr76$%^^" it accepts and registers.
5.This shouldn't happen. This is where the error lies.

**Expected behavior**
When tried to enter Name as "/*--*/+uioyr76$%^^" it should display a prompt message to enter a valid name and not let the user to sign up with invalid names.

**Device Information [optional]:**
 - OS: [windows 10]
 - Browser [chrome]
 - Version [Version 83.0.4103.106 (Official Build) (64-bit)]

I am working on this,
[name error.docx](https://github.com/CircuitVerse/CircuitVerse/files/4800361/name.error.docx)

### LINKED ISSUE #1633: user sign up, email field only checks for @ in the email.
The bug:
When a user signs up, if he leaves out the end part of the email, the user is still able to sign up without a proper mail id.

Example:
If a user tries to enter email as, user@test.com, the email format is correct.
but if he enters an email without the '@', the error message is displayed by the browser saying that the user is missing the '@'. 
The bug happens when a user enters email as user@test without any ending format, such as .com, .co, or .info, etc.
The user is still able to sign up.
**To Reproduce**

**Expected behavior**

**Screenshots**

**Device Information [optional]:**

**Additional context**

**Are you working on this issue? (Yes/No)**

### LINKED ISSUE #1640: Signup form improvements
**Is your feature request related to a problem? Please describe.**

Circuit verse signup forms has many validation problems 

problem1:no proper validation set up to username field

    as there is only one field in signup form it can used as username  as well as full name for user 
   so only possible validation can be applied is it cant be left blank  (Closes #1503 )
    (  no server side validation will be applicable as it will create conflicts with existing users)
                

problem2:  no proper validation set up in email field

                  *required type validation preexists on email field but it do not check the correctness
                    of the form of the email (except @ postion in email)

                  *you can put email without domain name of email  
 

**Describe the solution you'd like**
   I will apply each possible validations on email and username as follows:

    for username:
     * currently only validation applied is it cant be left blank
 
     for email:
     *we will check email in the form of

    1> local section of email
     2>@ symbol present
     3> domain name present (which takes only characters as a input)  (Closes #1633  )
   
    any of the part not present in email will consider email as invalid

**Are you working on this? (Yes/No)**
Yes.

### PR DESCRIPTION
**Pull request by pavanjoshi914**
closes #1640 ,closes #1503 ,closes #1633 ,closes #1649 

#### Describe the changes you have made in this PR -
1>adding client side validation 
2> updating devise regex with more powerfull regex
3>validating email with devise 
4>validating username with regex
5>fixing rspec tests and ensuring that factorybot generates the user which follow new validation rules
6>add specs for username validation
7> removing existing specs which conflicts client side validation

### Screenshots of the changes (If any) -

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

File: spec/system/signup_spec.rb [hunk 1]
```diff
@@ -11,40 +11,26 @@
...
-  it "should not sign-up when no credentials" do
-    click_button "Sign up"
-
-    expect(page).to have_text("Email can't be blank")
-    expect(page).to have_text("Password can't be blank")
-  end
-
-  it "should not sign-up when password is empty" do
-    fill_in "Name", with: "user1"
+  it "should not signup when username is invalid" do
```
[the remark this answers is not in the document]
Comment (pavanjoshi914) [reply spec/system/signup_spec.rb:14]: Not necessary making test describe in sync with others

**Comment by pavanjoshi914** [comment]
@satu0king @tachyons  please have a review on this!!

File: spec/system/signup_spec.rb [hunk 2]
```diff
@@ -11,48 +11,24 @@
...
   end
 
-  it "does not signup with special symbols" do
-    fill_in "Name", with: "!@#$%^&"
-    fill_in "Email", with: "user1@example.com"
-    fill_in "Password", with: "secret"
-    click_button "Sign up"
-
-    expect(page).to have_text("Name can only contain letters and spaces")
-  end
```
Comment (satu0king) [inline spec/system/signup_spec.rb:54]: Why is this test removed?

File: app/models/user.rb [hunk 3]
```diff
@@ -5,7 +5,10 @@ class User < ApplicationRecord
   require "pg_search"
   include SimpleDiscussion::ForumUser
 
-  validates :email, undisposable: { message: "Sorry, but we do not accept your mail provider." }
+  validates :email, format: { with: Devise.email_regexp },
+                    undisposable: { message: "Sorry, but we do not accept your mail provider." }
+  validates :name, format: { with: /\A[a-zA-Z ]{3,36}\z/,
```
Comment (satu0king) [inline app/models/user.rb:10]: New tests are needed to test if this functionality works as expected.

**Review by satu0king — CHANGES_REQUESTED** [review]

**Pull request closed without merging**